File: euler-problems/problems1-10/problem3/main.kt
import java.text.DecimalFormat
import kotlin.system.measureNanoTime

/**
 * The prime factors of 13195 are 5, 7, 13 and 29.
 * What is the largest prime factor of the number 600851475143 ?
 * Result: 6857
 */

fun solve1(): Int {
    var n = 600851475143L
    var i = 2

    while (n > 1) {
        while (n % i != 0L) {
            i++
        }
        n /= i
    }

    return i
}

fun solve2(): Int {
    tailrec fun helper(n: Long, i: Int): Int = when {
        n <= 1 -> i
        n % i != 0L -> helper(n, i + 1)
        else -> helper(n / i, i)
    }

    val n = 600851475143L
    return helper(n, 2)
}


fun main(args: Array<String>) {
    measureNanoTime { println(solve1()) }.let { println("solve1: ${DecimalFormat.getInstance().format(it)}") }
    measureNanoTime { println(solve2()) }.let { println("solve2: ${DecimalFormat.getInstance().format(it)}") }
}

File: euler-problems/problems1-10/problem3/main.dart

/**
 * The prime factors of 13195 are 5, 7, 13 and 29.
 * What is the largest prime factor of the number 600851475143 ?
 * Result: 6857
 */

int solve1() {
  int n = 600851475143;
  int i = 2;

  while (n > 1) {
    while (n % i != 0) {
      i++;
    }
    n ~/= i;
  }

  return i;
}

int _recursive(i, n) =>
    n <= 1 ? i : n % i != 0 ? _recursive(i + 1, n) : _recursive(i, n ~/ i);

int solve2() {
  int n = 600851475143;
  return _recursive(2, n);
}

main(List<String> arguments) {
  print(solve1());
  print(solve2());
}
